src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1293,34 +1310,47 @@ private void AbortStreams(int lastValidStream, Exception abortException)
 
                         _httpStreams.Remove(kvp.Value.StreamId);
                     }
-                    else if (NetEventSource.IsEnabled)
+                    else
                     {
-                        Trace($"Found {nameof(streamId)} {streamId} <= {lastValidStream}.");
+						if (NetEventSource.IsEnabled)
+	                    {
+	                        Trace($"Found {nameof(streamId)} {streamId} <= {lastValidStream}.");
+	                    }
+
+                        hasAnyActiveStream = true;
                     }
                 }
 
                 CheckForShutdown();
             }
 
-            if (shouldInvalidate)
+            if (!hasAnyActiveStream)
             {
-                // Invalidate outside of lock to avoid race with HttpPool Dispose()
-                // We should not try to grab pool lock while holding connection lock as on disposing pool,
-                // we could hold pool lock while trying to grab connection lock in Dispose().
-                _pool.InvalidateHttp2Connection(this);
+                if (!_disposed)
+                {
+                    _disposed = true;
+                    // Invalidate outside of lock to avoid race with HttpPool Dispose()
+                    // We should not try to grab pool lock while holding connection lock as on disposing pool,
+                    // we could hold pool lock while trying to grab connection lock in Dispose().
+                    _pool.InvalidateHttp2Connection(this);
```

Comment:
We still need to invalidate the connection even if we have active streams. The point of invalidating the connection is to remove it from the pool, so that subsequent requests will cause a new connection.

I'm surprised this passes tests as is... we should ensure we have a test that validates this. I.e. send GOAWAY with a pending request, issue a new request and ensure it creates a new connection, even without the old connection completing the original request yet. Do we have a test like this? If not, let's add one.

Reply:
BTW, I'm not really clear why we had the `shouldInvalidate` check before. It should always be safe to invalidate the connection, because the pool will check whether it's actually the connection in use currently or not. So if it's previously been invalidated already, it's a no-op.

Reply:
actually seems there is something wrong with the retry logic. I've tried couple of combinations of this and none of them works. I'll add disabled tests as part of this PR but will not try to address it here

Reply:
The tests you added are great for retry, thanks!

But I think there's another test we can add, which should pass: Have a pending request and send GOAWAY. Then after GOAWAY is received, issue a new request and make sure it comes in on a new connection. No retry needed.

The point here is to ensure we are invalidating the connection when we receive GOAWAY.

Reply:
@geoffkizer I already added GoAwayFrame_NewRequest_NewConnection as part of different issue, would that be sufficient?

Reply:
The problem I see with GoAwayFrame_NewRequest_NewConnection is that it completes the first response before sending the second request. Completing the first response will presumably cause the connection to be disposed and invalidated.

I want a test that would have caught the issue of not invalidating immediately on GOAWAY.

Reply:
I've updated https://github.com/dotnet/corefx/issues/34638 and moved it back to 3.0 since PINGs now work and it will be unblocked with this PR